### PR TITLE
Verilator coverage

### PR DESCRIPTION
**Pull request by yo96**
In this PR i added extra options in the `sverilog` backend for Verilator coverage. Peitian also fixed a caching issue of the translation pass.

### CODE REVIEW

**Comment by yo96** [comment]
I should probably add the same thing for the `yosys` backend as well.

**Comment by ptpan** [comment]
> I should probably add the same thing for the `yosys` backend as well.

Yeah my thought exactly

**Comment by jsn1993** [comment]
@ptpan I guess you should fix those conflicts instead of Yanghui

File: pymtl3/passes/sverilog/import_/ImportConfigs.py [hunk 1]
```diff
@@ -238,13 +247,18 @@ def create_vl_cmd( s ):
...
     warnings    = s.create_vl_warning_cmd()
 
     all_opts = [
       top_module, mk_dir, include, en_assert, opt_level, loop_unroll,
-      stmt_unroll, trace, warnings, flist, src
+      stmt_unroll, trace, warnings, flist, src, coverage, 
+      line_cov, toggle_cov,
     ]
 
+    print( f"verilator --cc {' '.join(opt for opt in all_opts if opt)}" )
```
Comment (jsn1993) [inline pymtl3/passes/sverilog/import_/ImportConfigs.py]: remove the print?

File: pymtl3/passes/sverilog/import_/ImportConfigs.py [hunk 2]
```diff
@@ -255,8 +269,13 @@ def create_cc_cmd( s ):
     c_src_files = " ".join(s.get_c_src_files())
     ld_flags = s.get_option("ld_flags")
     ld_libs = s.get_option("ld_libs")
+    coverage = "-DVM_COVERAGE" if s.get_option("coverage") or \
+                                  s.get_option("line_coverage") or \
+                                  s.get_option("toggle_coverage") else ""
+    print( f"g++ {c_flags} {c_include_path} {ld_flags}"
```
Comment (jsn1993) [inline pymtl3/passes/sverilog/import_/ImportConfigs.py]: remove this as well

Comment (yo96) [reply pymtl3/passes/sverilog/import_/ImportConfigs.py]: oops

File: pymtl3/passes/rtlir/util/utility.py [hunk 3]
```diff
@@ -36,3 +36,19 @@ def get_string( obj ):
...
+
+  upblks = m.get_update_blocks() - m.get_update_ff()
+  return [ x for x in m.get_update_block_order() if x in upblks ]
+
+def get_ordered_update_ff( m ):
+  """Return a list of update-ff update blocks that have deterministic order"""
+
+  return [ x for x in m.get_update_block_order() if x in m.get_update_ff() ]
+
+def get_ordered_all_upblks( m ):
```
Comment (jsn1993) [inline pymtl3/passes/rtlir/util/utility.py]: is this on purpose not to interleave update_ffs and updates?

Comment (ptpan) [reply pymtl3/passes/rtlir/util/utility.py]: I don’t think any file in this PR used this feature. But I think it might be useful to keep them apart.

Comment (jsn1993) [reply pymtl3/passes/rtlir/util/utility.py]: hmm can we remove it then?

**Comment by jsn1993** [comment]
Do you want to merge this in? @yo96

**Pull request merged**